library/core/src/option.rs

```diff
@@ -90,6 +90,205 @@
+//! | [`and_then`] | `Some(x)` | `x`            | `None`          | `None`    |
+//! | [`and_then`] | `Some(x)` | `x`            | `Some(y)`       | `Some(y)` |
+//! | [`filter`]   | `None`    | N/A            | (not evaluated) | `None`    |
+//! | [`filter`]   | `Some(x)` | `x`            | `false`         | `None`    |
+//! | [`filter`]   | `Some(x)` | `x`            | `true`          | `Some(x)` |
+//! | [`or_else`]  | `None`    | N/A            | `None`          | `None`    |
+//! | [`or_else`]  | `None`    | N/A            | `Some(y)`       | `Some(y)` |
+//! | [`or_else`]  | `Some(x)` | N/A            | (not evaluated) | `Some(x)` |
+//!
+//! [`and`]: Option::and
+//! [`and_then`]: Option::and_then
+//! [`filter`]: Option::filter
+//! [`or`]: Option::or
+//! [`or_else`]: Option::or_else
+//! [`xor`]: Option::xor
+//!
+//! ## Iterators
+//!
+//! An [`Option`] can be iterated over. This can be helpful if you need an
+//! iterator that is conditionally empty. The iterator will either produce
+//! a single value (when the [`Option`] is [`Some`]), or produce no values
+//! (when the [`Option`] is [`None`]). For example, [`into_iter`] acts like
+//! [`once(v)`] if the [`Option`] is [`Some(v)`], and like [`empty()`] if
+//! the [`Option`] is [`None`].
+//!
+//! Iterators over [`Option`] come in three types:
+//!
+//! * [`into_iter`] consumes the [`Option`] and produces the contained
+//!   value
+//! * [`iter`] produces an immutable reference of type `&T` to the
+//!   contained value
+//! * [`iter_mut`] produces a mutable reference of type `&mut T` to the
+//!   contained value
+//!
+//! [`Option`] implements the [`FromIterator`] trait, which allows an
+//! iterator over [`Option`] values to be collected into an [`Option`] of a
+//! collection of each contained value of the original [`Option`] values,
+//! or [`None`] if any of the elements was [`None`].
+//!
+//! [`empty()`]: crate::iter::empty
+//! [`FromIterator`]: Option#impl-FromIterator%3COption%3CA%3E%3E
+//! [`into_iter`]: Option::into_iter
+//! [`iter`]: Option::iter
+//! [`iter_mut`]: Option::iter_mut
+//! [`once(v)`]: crate::iter::once
+//! [`Some(v)`]: Some
+//!
+//! An iterator over [`Option`] can be useful when chaining iterators:
+//!
+//! ```
+//! let yep = Some(42);
+//! let nope = None;
+//! let nums: Vec<i32> = (0..4).chain(yep.into_iter()).chain(4..8).collect();
+//! assert_eq!(nums, [0, 1, 2, 3, 42, 4, 5, 6, 7]);
+//! let nums: Vec<i32> = (0..4).chain(nope.into_iter()).chain(4..8).collect();
+//! assert_eq!(nums, [0, 1, 2, 3, 4, 5, 6, 7]);
+//! ```
+//!
+//! One reason to chain iterators in this way is that a function returning
+//! `impl Iterator` must have all possible return values be of the same
+//! concrete type. Chaining an iterated [`Option`] can help with that.
+//!
+//! ```
+//! let yep = Some(42);
+//! let nope = None;
+//!
+//! fn makeiter(opt: Option<i32>) -> impl Iterator<Item = i32> {
+//!     (0..4).chain(opt.into_iter()).chain(4..8)
+//! }
```

Comment:
Hm, I think the reason that your proposed example happens to work is that the `None` gets inferred to be `Option<i32>` because the results of each arm of the `match` expression have to have identical types. This example was meant to be a stripped-down version of some other code I ran into, where there was a desire to return `impl Iterator` (instead of `Box<dyn Iterator>`), but there was difficulty constructing iterator chains with identical types based on some more complicated logic. Let me see if I can come up with a less-contrived example that is still succinct.